pisces/models.py

```diff
@@ -231,29 +234,29 @@ def prepare_set_for_training(self,
             with ProcessPoolExecutor(max_workers=workers_to_use) as executor:
                 results = list(
                     executor.map(
-                        self.get_needed_X_y_from_pair, 
-                        data_set_and_ids
+                        self.get_needed_X_y_from_tuple, 
+                        data_tuples
                     ))
         else:
             warnings.warn("No IDs found in the data set.")
             return results
         return results
     
-    def get_needed_X_y_from_pair(self, pair: Tuple[DataSetObject, str]) -> Tuple[np.ndarray, np.ndarray] | None:
+    def get_needed_X_y_from_tuple(self, data_tuple: Tuple[DataSetObject, str, bool]) -> Tuple[np.ndarray, np.ndarray] | None:
         """
         Get the needed X and y data from a pair of data set and ID.
 
         Args:
-            pair (Tuple[DataSetObject, str]): The pair of data set and ID.
+            data_tuple (Tuple[DataSetObject, str]): A tuple containing the data set, id, and N4 flag.
 
         Returns:
             Tuple[np.ndarray, np.ndarray] | None: The X and y data as a tuple, or None if an error occurred.
         """
-        data_set, id = pair
+        data_set, id, N4 = data_tuple
         print(f"getting needed X, y for {id}")
-        return self.get_needed_X_y(data_set, id)
+        return self.get_needed_X_y(data_set, id, N4)
```

Comment:
just a notational tip (which I didn't observe the first time writing this) we can actually do 
```python
return self.get_needed_X_y(*data_tuple)
```
which unpacks the tuple into ordered args.

Reply:
related tip, in cas you don't know: You can unpack a `dict` into the `kwargs` via `**dict_of_args`. This might actually be a better way to ensure this function is passing the correct args.

Instead of making tuples longer and longer, or pushing data into dicts with mysterious keys & values, what do you think of making a class:
```python
class DataSetSubject:
    def __init__(self, data_set: DataSetobject, id: str, N4: bool):
        self.data_set = data_set
        self.id = id
        self.N4 = N4   # remove if you put this info into DataSetObject
```
and modify this functions signature to:
```python
def get_needed_X_y_from_subject(self, subject: DataSetSubject) -> Tuple[np.ndarray, np.ndarray]:
    print(f"getting needed X, y for {subject.id}")
    return self.get_needed_X_y(**subject.__dict__)
```
We could even go a step further and make `get_needed_X_y(subject: DataSetSubject)` instead of `get_needed_X_y(data_set: DataSetObject, id: str, N4: bool)`

Adopting a pattern of using a class for complex, repeated arguments can also make it easier to refactor later, since we can just add or remove the `N4: bool` without changing a bunch of function signatures. Then apply the new info where it actually matters.

Reply:
I think this is a good approach and I implemented it this way! I'm thinking that in the future we should have `DataSetObject` be a collection of `DataSetSubject`s. We could have `N4: bool | List[bool]` as an input to `DataSetObject` where each individual could have its own `N4`. I tried implementing this but I realized it requires a big refactor of `DataSetObejct`, since it conflicts with how we find datasets and automatically create `DataSetObject`s for each. If we want to go this way it might be better to do it on another PR.